org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FormatterHandler.java

```diff
@@ -311,4 +313,44 @@ private IRegion getRegion(ICompilationUnit cu, IDocument document, Position posi
 		return null;
 	}
 
+	/**
+	 * A utility function to format a String with given formatter options.
+	 *
+	 * @see org.eclipse.jdt.core.formatter.CodeFormatterApplication#formatFile(File,
+	 *      CodeFormatter)
+	 * @param content
+	 *                    the content to format
+	 * @param options
+	 *                    the Map includes formatter options, If set to
+	 *                    <code>null</code>, then use the current settings from
+	 *                    <code>JavaCore#getOptions</code>.
+	 * @param version
+	 *                    the version of the formatter options
+	 * @param monitor
+	 *                    the progress monitor
+	 *
+	 * @return the formatted String
+	 */
+	public String stringFormatting(String content, Map<String, String> options, int version, IProgressMonitor monitor) {
+		IDocument document = new Document();
+		document.set(content);
+		if (options != null && version < ProfileVersionerCore.getCurrentVersion()) {

```

Comment:
In the LSP formatting request, we will merge the custom formatting options with the default options together before applying formatter. I suggest you to follow the similar logic.

Reply:
I've updated the PR to complete the logic when the options don't need to use `ProfileVersionCore.updateAndComplete()`